src/openms/include/OpenMS/COMPARISON/CLUSTERING/GridBasedClustering.h

```diff
@@ -244,32 +248,29 @@ namespace OpenMS
         grid_.removeCluster(cell_for_cluster2, cluster_index2);
         grid_.addCluster(cell_for_new_cluster, cluster_index1);
 
-        // update minimum distance list
-        std::vector<int> clusters_to_be_updated;
-        clusters_to_be_updated.push_back(cluster_index1);      // index of the new cluster
+        std::set<int> clusters_to_be_updated;
+        clusters_to_be_updated.insert(cluster_index1);
 
-        MultisetIterator it = distances_.begin();
-        while (it != distances_.end())
+        // erase distance object of cluster with cluster_index2 without updating (does not exist anymore!)
+        // (the one with cluster_index1 has already been erased at the top of the while loop)
+        eraseMinDistance_(distance_it_for_cluster_idx_[cluster_index2]);
+
+        // find out which clusters need to be updated
+        std::pair<NNIterator, NNIterator> nn_range = reverse_nns_.equal_range(cluster_index1);
+        for (NNIterator nn_it = nn_range.first; nn_it != nn_range.second;)
         {
-          MinimumDistance distance(*it);
-          if (distance.getClusterIndex() == cluster_index2)
-          {
-            // distance.getClusterIndex() == cluster_index1 should never happen since this entry has already been erased.
-            // Cluster 2 no longer exists. Hence only remove the entry from the list of minimum distances.
-            distances_.erase(it++);
-          }
-          else if (distance.getNearestNeighbourIndex() == cluster_index1 || distance.getNearestNeighbourIndex() == cluster_index2)
-          {
-            clusters_to_be_updated.push_back(distance.getClusterIndex());
-            distances_.erase(it++);
-          }
-          else
-          {
-            ++it;
-          }
+          clusters_to_be_updated.insert(nn_it->second->getClusterIndex());
+          eraseMinDistance_((nn_it++)->second);
+        }
+        nn_range = reverse_nns_.equal_range(cluster_index2);
+        for (NNIterator nn_it = nn_range.first; nn_it != nn_range.second;)
+        {
+          clusters_to_be_updated.insert(nn_it->second->getClusterIndex());
+          eraseMinDistance_((nn_it++)->second);
         }
 
-        for (std::vector<int>::const_iterator cluster_index = clusters_to_be_updated.begin(); cluster_index != clusters_to_be_updated.end(); ++cluster_index)
+        // update clusters
+        for (std::set<int>::const_iterator cluster_index = clusters_to_be_updated.begin(); cluster_index != clusters_to_be_updated.end(); ++cluster_index)
         {
           if (findNearestNeighbour_(clusters_.find(*cluster_index)->second, *cluster_index))
```

Comment:
With caching and some const references instead of copying it now takes 15 s instead of 18 s ;)